http/objectstore/iam/check_handler_auth.go

```diff
@@ -0,0 +1,289 @@
+package iam
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"github.com/filedag-project/filedag-storage/http/objectstore/api_errors"
+	"github.com/filedag-project/filedag-storage/http/objectstore/consts"
+	"github.com/filedag-project/filedag-storage/http/objectstore/iam/auth"
+	"github.com/filedag-project/filedag-storage/http/objectstore/iam/s3action"
+	"github.com/filedag-project/filedag-storage/http/objectstore/utils/etag"
+	"github.com/filedag-project/filedag-storage/http/objectstore/utils/hash"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+// AuthSys auth and sign system
+type AuthSys struct {
+	Iam       IdentityAMSys
+	PolicySys IPolicySys
+}
+
+//Init AuthSys
+func (s *AuthSys) Init() {
+	s.Iam.Init()
+	s.PolicySys.Init()
+}
+
+// CheckRequestAuthTypeCredential Check request auth type verifies the incoming http request
+// - validates the request signature
+// - validates the policy action if anonymous tests bucket policies if any,
+//   for authenticated requests validates IAM policies.
+// returns APIErrorCode if any to be replied to the client.
+// Additionally, returns the accessKey used in the request, and if this request is by an admin.
+func (s *AuthSys) CheckRequestAuthTypeCredential(ctx context.Context, r *http.Request, action s3action.Action, bucketName, objectName string) (cred auth.Credentials, owner bool, s3Err api_errors.ErrorCode) {
+	switch GetRequestAuthType(r) {
+	case authTypeUnknown, AuthTypeStreamingSigned:
+		return cred, owner, api_errors.ErrSignatureVersionNotSupported
+	case authTypePresignedV2, authTypeSignedV2:
+		if s3Err = s.isReqAuthenticatedV2(r); s3Err != api_errors.ErrNone {
+			return cred, owner, s3Err
+		}
+		cred, owner, s3Err = s.getReqAccessKeyV2(r)
+	case authTypeSigned, authTypePresigned:
+		region := ""
+		switch action {
+		case s3action.GetBucketLocationAction, s3action.ListAllMyBucketsAction:
+			region = ""
+		}
+		if s3Err = s.IsReqAuthenticated(ctx, r, region, serviceS3); s3Err != api_errors.ErrNone {
+			return cred, owner, s3Err
+		}
+		cred, owner, s3Err = s.GetReqAccessKeyV4(r, region, serviceS3)
+	}
+	if s3Err != api_errors.ErrNone {
+		return cred, owner, s3Err
+	}
+	if cred.IsTemp() {
+		cred, _ = s.Iam.GetUser(ctx, cred.ParentUser)
+	}
+	if action == s3action.CreateBucketAction {
+		// To extract region from XML in request body, get copy of request body.
+		payload, err := ioutil.ReadAll(io.LimitReader(r.Body, consts.MaxLocationConstraintSize))
+		if err != nil {
+			log.Errorf("ReadAll err:%v", err)
+			return cred, owner, api_errors.ErrMalformedXML
+		}
+
+		// Populate payload to extract location constraint.
+		r.Body = ioutil.NopCloser(bytes.NewReader(payload))
+
+		// Populate payload again to handle it in HTTP handler.
+		r.Body = ioutil.NopCloser(bytes.NewReader(payload))
```

Comment:
Is this duplicate code?

Reply:
Yes,between these two lines, there are some functions before, I forgot to delete the corresponding code when I delete these functions,I will fix it in next submission.